### PR TITLE
engine: Fix Cloe state machine

### PR DESCRIPTION
**Pull request by fgguidi**
Fix Cloe state machine for irregular situations to go through Abort and Disconnect.

### CODE REVIEW

**Comment by cassava** [comment]
So I spent some time looking through the code in `simulation.cpp`, and I think that the proposed solution is an improvement, but will not be very much more informative than the previous state. For example, I'm not sure that we'll know which model caused the abort/reset/error.

**Review by cassava — CHANGES_REQUESTED** [review]
I made some suggestions on what the logged messages should say. I think this is more clear, since we can also catch errors in the states themselves and deal better with them there.

File: engine/src/simulation.cpp [hunk 1]
```diff
@@ -199,6 +199,15 @@ class SimulationMachine
...
       } catch (cloe::AsyncAbort&) {
         this->push_interrupt(ABORT);
+      } catch (cloe::ModelReset& e) {
+        logger()->error("Resetting simulation due to error: {}", e.what());
+        this->push_interrupt(RESET);
+      } catch (cloe::ModelAbort& e) {
+        logger()->error("Aborting simulation due to error: {}", e.what());
+        this->push_interrupt(ABORT);
+      } catch (cloe::ModelError& e) {
+        logger()->critical("Aborting simulation due to error: {}", e.what());
```
[the remark this answers is not in the document]
Comment (fgguidi) [reply engine/src/simulation.cpp]: I have pushed your suggestion.

File: engine/src/simulation.cpp [hunk 2]
```diff
@@ -199,6 +199,15 @@ class SimulationMachine
         }
       } catch (cloe::AsyncAbort&) {
         this->push_interrupt(ABORT);
+      } catch (cloe::ModelReset& e) {
+        logger()->error("Resetting simulation due to error: {}", e.what());
```
[the remark this answers is not in the document]
Comment (fgguidi) [reply engine/src/simulation.cpp]: I have pushed your suggestion.

**Comment by fgguidi** [comment]
> I'm not sure that we'll know which model caused the abort/reset/error

**Comment by fgguidi** [comment]
> So I spent some time looking through the code in `simulation.cpp`, and I think that the proposed solution is an improvement, but will not be very much more informative than the previous state. For example, I'm not sure that we'll know which model caused the abort/reset/error.

@cassava  If you would like to be more informative then we should print more information in the different implementations of foreach_model() functions in simulation_context.cpp. Right?

**Comment by cassava** [comment]
> > So I spent some time looking through the code in `simulation.cpp`, and I think that the proposed solution is an improvement, but will not be very much more informative than the previous state. For example, I'm not sure that we'll know which model caused the abort/reset/error.
> 
> @cassava If you would like to be more informative then we should print more information in the different implementations of foreach_model() functions in simulation_context.cpp. Right?

Yes, that is correct. However, this was just an observation, I would just bear it in mind if we are dissatisfied with the output. Here I'd wait until we have actual examples. That can go in a separate PR.

**Review by cassava — APPROVED** [review]

**Pull request merged**